### PR TITLE
Fix stream trace logging with the OpenAI autologging implementation to record the correct chunk structure

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/12629?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12629/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12629
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Adjust the trace logging for streaming implementations with openai autologging to record the chunk structure rather than the stream object reference (current master branch logs the name of the openai.Stream object). 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests

Screenshot 2024-07-10 at 1 37 57 PM

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/helper_functions.py [hunk 1]
```diff
@@ -715,17 +715,20 @@ def start_mock_openai_server():
...
                     break
             else:
+                proc.kill()
+                proc.wait()
                 raise RuntimeError("Failed to start mock OpenAI server")
 
             yield base_url
         finally:
             proc.kill()
+            proc.wait()
```
Comment (harupy) [inline tests/helper_functions.py:734]: `wait` should be called when the context manager exits.

**Review by B-Step62 — APPROVED** [review]
LGTM. One random thought: The raw chunk list is hardly readable. What about storing the concatenated text in `output` and put chunks in event as auxiliary information? This is what LangSmith does for streaming (and what we do for LC). Counter argument might be that we avoid modifying output as much as possible.

**Comment by BenWilson2** [comment]
> LGTM. One random thought: The raw chunk list is hardly readable. What about storing the concatenated text in `output` and put chunks in event as auxiliary information? This is what LangSmith does for streaming (and what we do for LC). Counter argument might be that we avoid modifying output as much as possible.

I love this idea! Will make the change :)

File: tests/helper_functions.py (hunk 1 above)
Comment (BenWilson2) [reply tests/helper_functions.py:734]: I'll remove this

File: tests/helper_functions.py [hunk 2]
```diff
@@ -715,17 +715,20 @@ def start_mock_openai_server():
...
                     resp = requests.get(f"{base_url}/health")
                 except requests.ConnectionError:
-                    time.sleep(1)
+                    time.sleep(2)
                     continue
                 if resp.ok:
                     break
             else:
+                proc.kill()
+                proc.wait()
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply tests/helper_functions.py:728]: When running this suite locally, I was getting zombie processes when a test failed that required an fkill of the python kernel to run again. 
The other modifications (increasing timeout for health check were to get a more repeatable non-flaky local execution of the suite).

[the remark this answers is not in the document]
Comment (harupy) [reply tests/helper_functions.py:728]: @BenWilson2 interesting. kill In finally isn't called?

[the remark this answers is not in the document]
Comment (harupy) [reply tests/helper_functions.py:728]: I made this change:

```diff
diff --git a/tests/helper_functions.py b/tests/helper_functions.py
index 616b66fd5..546a818bb 100644
--- a/tests/helper_functions.py
+++ b/tests/helper_functions.py
@@ -714,6 +714,7 @@ def start_mock_openai_server():
         ]
     ) as proc:
         try:
+            print("Started server")
             base_url = f"http://localhost:{port}"
             for _ in range(10):
                 try:
@@ -724,10 +725,11 @@ def start_mock_openai_server():
                 if resp.ok:
                     break
             else:
-                proc.kill()
-                proc.wait()
+                # proc.kill()
+                # proc.wait()
                 raise RuntimeError("Failed to start mock OpenAI server")
 
             yield base_url
         finally:
+            print("Killing server")
             proc.kill()
diff --git a/tests/openai/test_openai_autolog.py b/tests/openai/test_openai_autolog.py
index ccb747023..4e92b51b4 100644
--- a/tests/openai/test_openai_autolog.py
+++ b/tests/openai/test_openai_autolog.py
@@ -126,6 +126,7 @@ def test_completions_autolog_artifacts(client, monkeypatch):
 
 @pytest.mark.skipif(not is_v1, reason="Requires OpenAI SDK v1")
 def test_completions_autolog_streaming(client, monkeypatch):
+    raise Exception("test failed")
     mlflow.openai.autolog(log_models=True)
     with mlflow.start_run() as run:
         client.completions._mlflow_session_id = "test_session_id"
```

and ran:

```
pytest tests/openai/test_openai_autolog.py::test_completions_autolog_streaming -s
```

but could not repro. Can you try?

[the remark this answers is not in the document]
Comment (BenWilson2) [reply tests/helper_functions.py:728]: Just tried it and it hangs on my machine. SUPER strange.

[the remark this answers is not in the document]
Comment (BenWilson2) [reply tests/helper_functions.py:728]: I'll do a full python reinstall tomorrow and if it works correctly then, I'll remove the redundant calls.

[the remark this answers is not in the document]
Comment (harupy) [reply tests/helper_functions.py:728]: @BenWilson2 Please don't try to reinstall python :) Let's look into what's hanging.

**Pull request merged**